iocore/utils/SharedExtendible.h

```diff
@@ -0,0 +1,447 @@
+      return &fields[field_name].offset;
+    }
+
+    FieldId_c
+    find_c(char const *field_name)
+    {
+      auto field_iter = fields.find(field_name);
+      ink_release_assert(field_iter != fields.end());
+      FieldSchema &fs = field_iter->second;
+      ink_release_assert(fs.access == C_API);
+      return &fs.offset;
+    }
+
+    void
+    updateMemOffsets()
+    {
+      ink_release_assert(instance_count == 0);
+
+      uint32_t acc_offset = ROUNDUP(sizeof(Derived_t), alloc_align);
+
+      SharedOffset_t size_blocks[] = {4, 2, 1};
+
+      for (auto &pair_fld : fields) {
+        auto &fld = pair_fld.second;
+        if (fld.size >= 8) {
+          fld.offset = acc_offset;
+          acc_offset += fld.size;
+        }
+      }
+      for (auto sz : size_blocks) {
+        for (auto &pair_fld : fields) {
+          auto &fld = pair_fld.second;
+          if (fld.size == sz) {
+            fld.offset = acc_offset;
+            acc_offset += fld.size;
+          }
+        }
+      }
+      bit_offset              = acc_offset;
+      uint32_t acc_bit_offset = 0;
+      for (auto &pair_fld : fields) {
+        auto &fld = pair_fld.second;
+        if (fld.size == 0) {
+          fld.offset = acc_bit_offset;
+          ++acc_bit_offset;
+        }
+      }
+
+      alloc_size = acc_offset + (acc_bit_offset + 7) / 8; // size '0' are packed bit allocations.
+    }
+
+    bool
+    reset()
+    {
+      if (instance_count > 0) {
+        // free instances before calling this so we don't leak memory
+        return false;
+      }
+      fields.clear();
+      alloc_size  = sizeof(Derived_t);
+      alloc_align = alignof(Derived_t);
+      return true;
+    }
+
+    void
+    call_construct(char *ext_as_char_ptr)
+    {
+      ++instance_count; // don't allow schema modification
+      // init all extendible memory to 0, incase constructors don't
+      memset(ext_as_char_ptr + sizeof(Derived_t), 0, alloc_size - sizeof(Derived_t));
+
+      for (auto const &elm : fields) {
+        FieldSchema const &field_schema = elm.second;
+        if (field_schema.access != BIT && field_schema.construct_fn != nullptr) {
+          field_schema.construct_fn(ext_as_char_ptr + field_schema.offset);
+        }
+      }
+    }
+
+    void
+    call_destruct(char *ext_as_char_ptr)
+    {
+      for (auto const &elm : fields) {
+        FieldSchema const &field_schema = elm.second;
+        if (field_schema.access != BIT && field_schema.destruct_fn != nullptr) {
+          field_schema.destruct_fn(ext_as_char_ptr + field_schema.offset);
+        }
+      }
+      --instance_count;
+    }
+    size_t
+    size() const
+    {
+      return alloc_size;
+    }
+    bool
+    no_instances() const
+    {
+      return instance_count == 0;
+    }
+
+  }; // end Schema struct
+
+  //////////////////////////////////////////
+  // SharedExtendible static data
+  /// one schema instance per Derived_t to define contained fields
+  static Schema schema;
```

Comment:
This is a header file. Won't this put a separate instance of `Schema` in every compilation unit that includes this header?

Reply:
that is within the SharedExtendible template. as with 98% of this file.